File: question/neetcode150/sliding_window/longest-repeating-character-replacement.py
from collections import defaultdict

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        d = defaultdict(int)

        l = 0
        m = 0
        for r in range(len(s)):
            d[s[r]] += 1
            m = max(m, d[s[r]])

            if r-l+1 > k + m:
                d[s[l]] -= 1
                l+=1
        
        return m

File: question/neetcode150/sliding_window/longest-repeating-character-replacement.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        int l = 0;
        int n = s.size();
        int m = 0;
        int ans = 0;

        unordered_map<char, int> d;

        for (int r = 0;r < n;r++){
            char c = s[r];
            d[c]++;
            m = max(m, d[c]);

            if(r-l+1 > m+k){
                d[s[l]]--;
                l++;
            }

            ans = max(ans, r-l+1);
        }

        return ans;
    }
};